### DISCUSSION TITLE
Ideas to avoid zha/zigpy/HW overload

### DISCUSSION THREAD

**Discussion by mdeweerd** [zigpy/zigpy#1049] [ZHA development]
I am not sure where to best post this as ZHA and zigpy are closely related.

My network was slow/not responding and I had a closer look at what was happening (shared further below), but first my ideas based on that:

For the record, I was on:
Home Assistant 2022.9.4
Supervisor 2022.09.1
Operating System 8.4
- Avoid requesting non-existing attributes (monitor for attributes for which the read requests resulted in the error(s) that they do not exist);
- Request attributes only if they are old (reportable attributes do not need to be requested if they have been reported no later than the maximum reportable time (+ 10 seconds of margin for instance));
- Provide a means to adjust the attributes that will be requested 
- Provide a means to restart ZHA/ZIGPY with a reset of the key (given that an automation could be set up to monitor for a given number of errors over a given time period, and reset zha/zigpy when above a limit)
- Internally monitor the error rate (zha/zigpy) to reset the component.

# Experience/observations

When I start writing this, my ZHA network is responding slowly, so I enabled the debug log level for zha givent the errors observed in the log.

```
2022-09-20 18:55:06.377 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/zigpy_znp/api.py", line 1025, in request_callback_rsp
    return await callback_rsp
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/zigpy_znp/api.py", line 1022, in request_callback_rsp
    async with async_timeout.timeout(timeout):
  File "/usr/local/lib/python3.10/site-packages/async_timeout/__init__.py", line 129, in __aexit__
    self._do_exit(exc_type)
  File "/usr/local/lib/python3.10/site-packages/async_timeout/__init__.py", line 212, in _do_exit
    raise asyncio.TimeoutError
asyncio.exceptions.TimeoutError
2022-09-20 18:55:13.174 DEBUG (MainThread) [homeassistant.components.zha.core.device] [0x78E4](01MINIZB): Device seen - marking the device available and resetting counter
2022-09-20 18:55:13.175 DEBUG (MainThread) [homeassistant.components.zha.core.device] [0x78E4](01MINIZB): Update device availability -  device available: True - new availability: True - changed: False
2022-09-20 18:55:13.290 DEBUG (MainThread) [homeassistant.components.zha.entity] light.tz3000_dbou1ap4_ts0505a_9afc91fe_level_light_color_on_off: setting transitioning flag to True
2022-09-20 18:55:14.369 DEBUG (MainThread) [homeassistant.components.zha.core.channels.base] [0xBC00:1:0x0b04]: failed to get attributes '['active_power', 'active_power_max', 'apparent_power', 'rms_current', 'rms_current_max']' on 'electrical_measurement' cluster:
2022-09-20 18:55:14.370 DEBUG (MainThread) [homeassistant.components.zha.core.channels.base] [0xBC00:1:0x0b04]: Reading attributes in chunks: ['rms_voltage']
2022-09-20 18:55:16.352 WARNING (MainThread) [homeassistant.helpers.entity] Update of sensor.lixee_zlinky_tic_ea99d205_electrical_measurement is taking over 10 seconds
```

Which suggests that ['rms_voltage'] and '['active_power', 'active_power_max', 'apparent_power', 'rms_current', 'rms_current_max']' are fetched at about the same time.
That helps loading the zigbee network.

AFAICS my device does not have active_power values, only apparent_power, so thos values could be avoided and therefore have only one request.  Two might be too much for my device.
At least, I would suggest to leave more time between two requests.

I also got quite a few MEM_ERRORs:
```
2022-09-20 19:10:22.015 DEBUG (MainThread) [homeassistant.components.zha.core.channels.base] [0xBC00:1:0x0b04]: failed to get attributes '['rms_voltage']' on 'electrical_measurement' cluster: Request failed after 5 attempts: <Status.MEM_ERROR: 16>
```

Which suggest that the CC2531 device does not have enough memory to cope.  Reducing the number of requests could help, and maybe even deciding to reset the Coordinator to clean its slate may be usefull.

This also happens during a backup:
```
2022-09-20 15:20:37.822 WARNING (MainThread) [zigpy.backups] Failed to create a network backup
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/zigpy_znp/nvram.py", line 177, in osal_read
    read_rsp = await self.znp.request(
  File "/usr/local/lib/python3.10/site-packages/zigpy_znp/api.py", line 993, in request
    raise InvalidCommandResponse(
zigpy_znp.exceptions.InvalidCommandResponse: Expected SRSP response SYS.OSALNVReadExt.Rsp(Status=<Status.SUCCESS: 0>, Value=None), got SYS.OSALNVReadExt.Rsp(Status=<Status.MEM_ERROR: 16>, Value=b'')
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/zigpy/backups.py", line 207, in _backup_loop
    await self.create_backup()
  File "/usr/local/lib/python3.10/site-packages/zigpy/backups.py", line 128, in create_backup
    await self.app.load_network_info(load_devices=load_devices)
  File "/usr/local/lib/python3.10/site-packages/zigpy_znp/zigbee/application.py", line 164, in load_network_info
    await self._znp.load_network_info(load_devices=load_devices)
  File "/usr/local/lib/python3.10/site-packages/zigpy_znp/api.py", line 107, in load_network_info
    nib = await self.nvram.osal_read(OsalNvIds.NIB, item_type=t.NIB)
  File "/usr/local/lib/python3.10/site-packages/zigpy_znp/nvram.py", line 185, in osal_read
    assert e.response.Status == t.Status.INVALID_PARAMETER
AssertionError
```

I used to do my backups during the night and while there are few Zigbee transactions, so maybe that is one of the contributers for seeing the slowness (and absence of reaction) now.

I removed and reinserted the CC2531 key and the behavior became normal again.

All in all I think this has been going on for about a day - I was wonderning why my movement controlled light was still on when I made a nightly "pit stop", and my logs showed al lot of the errors occuring since about 1 day.

**Comment by puddly** [comment]
CC2531 support at this point is best-effort: its firmware has been end-of-life for more than a few years now.  Use the new Zigbee migration feature in ZHA (it also performs network backups every 24 hours automatically) to migrate to a CC2652 or a recent EmberZNet coordinator.

That being said, what firmware are you running on your CC2531? I haven't encountered `MEM_ERROR` during NVRAM reads before.

**Reply by mdeweerd** [reply]
I am not sure when I last flashed it - but I think it is the latest one (ZStack 3.0.x since the last change was 3 years ago (https://github.com/Koenkk/Z-Stack-firmware/tree/master/coordinator/Z-Stack_3.0.x/bin ).

Regarding the migration: I am holding off a bit until there has been confirmed feedbacks - I do not have a lot of time on my hands to handle the case where I need to cope with "exceptions".

**Reply by puddly** [reply]
Downgrade to Z-Stack Home 1.2 (or switch adapters!). Z-Stack 3.0.x is unstable on the CC2531.

**Reply by Hedda** [reply]
Highly recommend migrating to a TI CC2652P based adapter or Silicon Labs EFR32MG21 based adapter. Those two feature ”modern” specifications and power amplification with configurable transmission power function

**Reply by puddly** [reply]
> Regarding the migration: I am holding off a bit until there has been confirmed feedbacks - I do not have a lot of time on my hands to handle the case where I need to cope with "exceptions".

The CC2531 -> CC2652 was the first migration I wrote for zigpy-znp so it's arguably the most well-tested at this point. If you want to avoid the new ZHA config flow, migrate manually with zigpy CLI and fix the serial port path in `/config/.storage/core.config_entries`.

**Comment by mdeweerd** [comment]
I was planning on buying and upgrading once I could expect to have more time on my hands.

However, I do think that (at least some of) my proposals are valid regardless of the fact that I am using a CC2351.

**Reply by Hedda** [reply]
How many devices are connected to it directly (direct children), how many devices are connected indirectly through router devices, and how many of those type of connected devices are Zigbee 3.0 devices versus how many are Zigbee Home Automation / Zigbee Light Link devices? 

FYI, here are by the way some community references that mention some different limitations, etc. related to CC235x as Zigbee Coordinator:

https://github.com/Koenkk/Z-Stack-firmware/blob/master/coordinator/README.md#overiew

https://github.com/Koenkk/Z-Stack-firmware/blob/master/coordinator/README.md#im-using-a-cc2530-or-cc2531-which-firmware-should-i-use

https://www.home-assistant.io/integrations/zha#known-working-zigbee-radio-modules

https://github.com/zigpy/zigpy-znp/blob/dev/README.md#hardware-requirements

https://www.zigbee2mqtt.io/guide/adapters/#not-recommended

https://www.zigbee2mqtt.io/guide/faq/#i-read-that-zigbee2mqtt-has-a-limit-of-20-devices-when-using-a-cc2530-cc2531-adapter-is-this-true

**Reply by mdeweerd** [reply]
IHMO the CC253x does not take more direct children and indirect children than it can add to its internal database.  It's a firmware configuration that has been optimized by Koenk by deactivating several other functions that are enabled in the default compilation settings.

I understand that the CC253x is limited, but that does not mean that the lessons learnt are to be ignored.

> Avoid requesting non-existing attributes (monitor for attributes for which the read requests resulted in the error(s) that they do not exist).

Requesting attributes that do not exist just generates extra traffic on the network and more effort from the target device.  The case I observed is a "permanently powered" device, but it is powered from a "low power source".   That is: it is powered by a specific output of the electrical meter which has a relatively high output impedance to limit the (free) power you can get from it.
According to the designer it can't cope with a lot of requests (replies in particular) because it breaks down the reserve power.

In this case because the number of attributes requested exceeds 5, there are two zigbee requests that are made in stead of one.

> Request attributes only if they are old (reportable attributes do not need to be requested if they have been reported no later than the maximum reportable time (+ 10 seconds of margin for instance));

One purpose of having reportable attributes is to avoid the traffic from requests and "useless" requests, while keeping up to date.  It is also a method that allows the end device to decide for itself when it is best to send the data (in light of power management for instance).

> Provide a means to adjust the attributes that will be requested

This would allow for adjusting for usefull attributes to request, either through a quirk configuration, or because the user trims or extends the attributes to those that (s)he's really interested in.

> Provide a means to restart ZHA/ZIGPY with a reset of the key (given that an automation could be set up to monitor for a given number of errors over a given time period, and reset zha/zigpy when above a limit).

Regardless of the device things can go wrong, and there is already some provisioning to detect that the serial connection broke down and restart.
Extending that feature (reset/restart of the connection) to be triggered by a monitoring condition failure helps keeping the system survive sporadic failures.
Making the reset available to an automation provides flexibility to the "community".

> Internally monitor the error rate (zha/zigpy) to reset the component.

This would make it easier on the community as implementing/copying automations that do the monitoring is not easily achievable for everybody.

**Reply by mdeweerd** [reply]
@puddly I just migrated  CC2531 to Soniff/ITEAD.  Worked fast and well.

I would just recommend indicating that a backup is made before erasing the original key.

**Reply by Hedda** [reply]
> I would just recommend indicating that a backup is made before erasing the original key.

+1 and maybe mention potential consequences of massive network degradation if the old adapter is not reset and still powered on.

https://github.com/home-assistant/core/pull/79663

> _Dual-radio adapters like the HUSBZB-1 cannot be unplugged after migration: **this causes two concurrent Zigbee networks with the same settings to be active, causing massive network degradation on the new radio.**_


https://github.com/home-assistant/core/blob/dev/homeassistant/components/zha/strings.json#L87-L90

```
      "intent_migrate": {
        "title": "Migrate to a new radio",
        "description": "Your old radio will be factory reset.  If you are using a combined Z-Wave and Zigbee adapter like the HUSBZB-1, this will only reset the Zigbee portion.\n\nDo you wish to continue?"
      },
```

![image](https://user-images.githubusercontent.com/6320001/194812705-4339f16d-2605-4d54-a13d-8d8a8cc84194.png)
